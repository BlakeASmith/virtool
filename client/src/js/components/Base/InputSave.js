/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports InputSave
 */

import React from "react";
import PropTypes from "prop-types";
import { Flex, FlexItem, Input, Button } from "./";

/**
 * A single input form component with a submit addon button that behaves well for updated VT settings.
 */
export class InputSave extends React.Component {

    constructor (props) {
        super(props);

        this.state = {
            value: this.props.initialValue,
            pending: false
        };
    }

    static propTypes = {
        name: PropTypes.string,

        // The function to call with the newValue when the form is submitted.
        onSave: PropTypes.func.isRequired,

        // The label to apply to the Input component.
        label: PropTypes.string,

        // The type attribute to apply to the Input component.
        type: PropTypes.string,

        // The value that should initially be given to the Input component.
        initialValue: PropTypes.any,

        // The autocomplete attribute to be passed to the
        autoComplete: PropTypes.bool,

        disabled: PropTypes.bool
    };

    static defaultProps = {
        initialValue: "",
        autoComplete: true
    };

    componentWillReceiveProps (nextProps) {
        // If the initialValue has changed. Remove the pending state on the component. This will remove the spinner on
        // the save button and enable the Input component again.
        if (nextProps.initialValue !== this.props.initialValue) {
            this.setState({
                pending: false,
                value: nextProps.initialValue
            });
        }
    }

    /**
     * Resets the setting value to the initialValue if the form component loses focus. Clicking the saveButton does not
     * make the form lose focus. The form is intentionally blurred once an updated initialValue is received in props.
     */
    handleBlur = (event) => {
        if (!this.state.pending && event.relatedTarget && event.relatedTarget.type !== "submit") {
            this.setState({value: this.props.initialValue});
        }
    };

    /**
     * Handle the data from a change in the input element. Updates state to reflect what is being typed by the user.
     *
     * @param event {event} - the change event from the FormControl
     */
    handleChange = (event) => {
        this.setState({value: event.target.value});
    };

    /**
     * Handles a submit event on the form. If the setting has changed the onSave function (passed as a prop) is called.
     * A spinner is shown to indicate the change is pending. The spinner will be removed when a new initialValue is
     * received in props.
     *
     * @param event {object} - the submit event.
     */
    handleSubmit = (event) => {
        event.preventDefault();

        if (this.state.value !== this.props.initialValue) {
            // If the new value is different to the initial one, show a spinner and call the onSave function. Show a
            // spinner to indicate the request is pending. Drop focus from the form children.
            this.setState({pending: true}, () => {
                this.props.onSave({
                    value: this.state.value,
                    name: this.props.name
                });

                this.blur();
            });
        } else {
            // Drop focus from the form children even though no information has been sent to the server or passed to the
            // parent component.
            this.blur();
        }
    };

    focus = () => {
        this.inputNode.focus();
    };

    /**
     * Blur all focus-sensitive elements in the component.
     *
     * @func
     */
    blur = () => {
        this.inputNode.blur();
        this.buttonNode.blur();
    };

    render () {

        const disabled = this.state.pending || this.props.disabled;

        return (
            <form onSubmit={this.handleSubmit}>
                <h5><strong>{this.props.label}</strong></h5>
                <Flex alignItems="stretch" style={{marginBottom: "15px"}}>
                    <FlexItem grow={1} shrink={0}>
                        <Input
                            ref={(node) => this.inputNode = node}
                            type={this.props.type}
                            autoComplete={this.props.autoComplete}
                            onChange={this.handleChange}
                            onBlur={this.handleBlur}
                            value={this.state.value}
                            disabled={disabled}
                            noMargin
                        />
                    </FlexItem>
                    <Button
                        type="submit"
                        bsStyle="primary"
                        disabled={disabled}
                        icon="floppy"
                        ref={(button) => this.buttonNode = button}
                    />
                </Flex>
            </form>
        );


    }

}

