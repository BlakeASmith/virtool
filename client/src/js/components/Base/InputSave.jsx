/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports InputSave
 */

'use strict';

var React = require('react');
var ReactDOM = require('react-dom');
var Input = require('react-bootstrap/lib/Input');
var Button = require('react-bootstrap/lib/Button');
var Icon = require('virtool/js/components/Base/Icon.jsx');

/**
 * A single input form component with a submit addon button that behaves well for updated VT settings.
 */
var InputSave = React.createClass({

    propTypes: {
        name: React.PropTypes.string,

        // The function to call with the newValue when the form is submitted.
        onSave: React.PropTypes.func.isRequired,

        // The label to apply to the Input component.
        label: React.PropTypes.string,

        // The type attribute to apply to the Input component.
        type: React.PropTypes.string,

        // The value that should initially be given to the Input component.
        initialValue: React.PropTypes.any,

        // The autocomplete attribute to be passed to the
        autoComplete: React.PropTypes.string,

        disabled: React.PropTypes.bool
    },

    getDefaultProps: function () {
        return {
            type: 'text',
            initialValue: '',
            autoComplete: 'off'
        };
    },

    getInitialState: function () {
        return {
            value: this.props.initialValue,
            pending: false
        };
    },

    componentWillReceiveProps: function (nextProps) {
        // If the initialValue has changed. Remove the pending state on the component. This will remove the spinner on
        // the save button and enable the Input component again.
        if (nextProps.initialValue !== this.props.initialValue) {
            this.setState({
                pending: false,
                value: nextProps.initialValue
            });
        }
    },

    /**
     * Resets the setting value to the initialValue if the form component loses focus. Clicking the saveButton does not
     * make the form lose focus. The form is intentionally blurred once an updated initialValue is receieved in props.
     */
    handleBlur: function (event) {
        var focusedOnButton = event.relatedTarget && event.relatedTarget.type !== 'submit';
        if (!this.state.pending && focusedOnButton) this.setState({value: this.props.initialValue});
    },

    /**
     * Handles a change in the input element. Updates state to reflect what is being typed by the user.
     *
     * @param event {object} - the change event.
     */
    handleChange: function (event) {
        this.setState({value: event.target.value});
    },

    /**
     * Handles a submit event on the form. If the setting has changed the onSave function (passed as a prop) is called.
     * A spinner is shown to indicate the change is pending. The spinner will be removed when a new initialValue is
     * received in props.
     *
     * @param event {object} - the submit event.
     */
    handleSubmit: function (event) {
        event.preventDefault();

        if (this.state.value !== this.props.initialValue) {
            // If the new value is different to the initial one, show a spinner and call the onSave function. Show a
            // spinner to indicate the request is pending. Drop focus from the form children.
            this.setState({pending: true}, function () {
                this.props.onSave({value: this.state.value, name: this.props.name});
                this.removeFocus();
            });
        } else {
            // Drop focus from the form children even though no information has been sent to the server or passed to the
            // parent component.
            this.removeFocus();
        }
    },

    /**
     * Blur all focusable elements in the component.
     *
     * @func
     */
    removeFocus: function () {
        ReactDOM.findDOMNode(this.refs.button).blur();
        this.refs.input.getInputDOMNode().blur();
    },

    render: function () {
        // Show a warning if the value for autoComplete to be passed to the Input component is illegal.
        if (['off', 'on'].indexOf(this.props.autoComplete) === -1) {
            console.warn('Illegal value for InputSave autoComplete prop');
        }

        // The addon button to attach to the Input component.
        var saveButton = (
            <Button ref='button' bsStyle='primary' type='submit' disabled={this.props.disabled}>
                <Icon name='floppy' pending={this.state.pending} />
            </Button>
        );

        return (
            <form onSubmit={this.handleSubmit}>
                <Input
                    name={this.props.name}
                    type={this.props.type}
                    autoComplete={this.props.autoComplete}
                    label={this.props.label}
                    onChange={this.handleChange}
                    onBlur={this.handleBlur}
                    ref='input'
                    buttonAfter={saveButton}
                    value={this.state.value}
                    disabled={this.state.pending || this.props.disabled}
                />
            </form>
        );
    }

});

module.exports = InputSave;
