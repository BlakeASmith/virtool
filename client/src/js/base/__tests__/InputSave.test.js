import { FormControl } from "react-bootstrap";
import { InputSave } from "../InputSave";
import { Flex, FlexItem, Button } from "../index";

describe("<InputSave />", () => {
    let props;
    let state;
    const blur = jest.fn();
    const e = {
        target: {
            value: "foo"
        },
        preventDefault: jest.fn()
    };
    beforeEach(() => {
        props = {
            name: "test_inputsave",
            onSave: jest.fn(),
            label: "Test Label",
            type: "number",
            min: 0,
            max: 10,
            step: 1,
            initialValue: 5,
            autoComplete: false,
            onInvalid: jest.fn(),
            onChange: jest.fn(),
            onBlur: jest.fn(),
            disabled: false,
            noMargin: true,
            error: "test error"
        };
        state = {
            pending: false,
            value: 5
        };
    });

    it("should render", () => {
        const wrapper = shallow(<InputSave {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("componentDidUpdate() should update state when [this.props.initialValue!=prevProps.initialValue]", () => {
        const wrapper = shallow(<InputSave {...props} />);
        expect(wrapper.state().value).toBe(5);
        wrapper.setState({
            pending: true,
            value: 8
        });
        wrapper.setProps({
            initialValue: 9
        });
        expect(wrapper.state()).toEqual({ pending: false, value: 9 });
    });

    it("should not update state when [this.props.initialValue=prevProps.initialValue]", () => {
        const wrapper = shallow(<InputSave {...props} />);
        expect(wrapper.state()).toEqual({ pending: false, value: 5 });
    });

    it("handleBlur should update state when [this.props.disable=false]", () => {
        const wrapper = mount(<InputSave {...props} />);
        wrapper.setState({ value: 6 });
        wrapper.find("FormControl").simulate("blur");
        expect(wrapper.state().value).toBe(5);
    });

    it("handleChange should update state when [this.props.disabled=false]]", () => {
        const wrapper = shallow(<InputSave {...props} />);
        wrapper.find("FormControl").simulate("change", e);
        expect(wrapper.state()).toEqual({ pending: false, value: "foo" });
    });

    it("handleChange should not update state when [this.props.disabled=true]]", () => {
        props.disabled = true;
        const wrapper = shallow(<InputSave {...props} />);
        wrapper.find("FormControl").simulate("change", e);
        expect(wrapper.state()).toEqual({ pending: false, value: 5 });
    });

    it("handleSubmit should call props.onSave when [this.state.value!=this.props.initialValue]", () => {
        const wrapper = mount(<InputSave {...props} />);
        wrapper.setState({ ...state, value: 10 });
        wrapper.find("FormControl").simulate("submit", e);
        expect(props.onSave).toHaveBeenCalledWith({
            max: 10,
            min: 0,
            name: "test_inputsave",
            value: 10
        });
    });

    it("handleSubmit should not call props.onSave when [this.state.value=this.props.initialValue]", () => {
        const wrapper = mount(<InputSave {...props} />);
        wrapper.find("FormControl").simulate("submit", e);
        expect(props.onSave).not.toHaveBeenCalled();
    });

    it("handleSubmit should not call props.onSave when [this.state.value=this.props.initialValue]", () => {
        const wrapper = mount(<InputSave {...props} />);
        wrapper.find("FormControl").simulate("invalid");
        expect(props.onInvalid).toHaveBeenCalled();
    });
});
