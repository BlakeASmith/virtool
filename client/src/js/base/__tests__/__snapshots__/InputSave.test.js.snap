// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputSave /> should render 1`] = `
<form
  onSubmit={[Function]}
>
  <h5>
    <strong>
      Test Label
    </strong>
  </h5>
  <Flex
    alignContent="stretch"
    alignItems="stretch"
    direction="row"
    justifyContent="flex-start"
    style={
      Object {
        "marginBottom": "0",
      }
    }
    wrap="nowrap"
  >
    <FlexItem
      alignSelf={null}
      basis="auto"
      grow={1}
      shrink={1}
    >
      <FormControl
        autoComplete="off"
        bsClass="form-control"
        className="form-control-error"
        componentClass="input"
        disabled={false}
        inputRef={[Function]}
        max={10}
        min={0}
        name="test_inputsave"
        onBlur={[Function]}
        onChange={[Function]}
        onInvalid={[MockFunction]}
        step={1}
        style={
          Object {
            "marginBottom": "0",
          }
        }
        type="number"
        value={5}
      />
    </FlexItem>
    <Button
      bsStyle="primary"
      disabled={false}
      icon="save"
      pullRight={false}
      tipPlacement="top"
      type="submit"
    />
  </Flex>
</form>
`;
